### ISSUE TITLE
feat: add option to stop validation after first validation error

### ISSUE THREAD

**Issue by HonoluluHenk** [typestack/class-validator#188]
If i apply the following annotations:

```ts
  @IsNotEmpty()
  @IsInt()
  @Min(0)
  foobar: number;
```

The validation result will be:

```json
{ 
  "value": null, 
  "property": "foobar", 
  "children": [], 
  "constraints": {
    "min": "foobar must be greater than 0",
   "isInt": "foobar must be an integer number", 
   "isNotEmpty": "foobar should not be empty" 
  } 
}
```

If no value is given (i.e.: `foobar === null`), I only want the validation result of `@IsNotEmpty`.
All the others are quite useless... worse: since the validation errors (the constraints) are stored in an object, the ordering is lost so I cannot "just take the first error constraint" :(

Expected result:
```
{ 
  "value": null, 
  "property": "foobar", 
  "children": [], 
  "constraints": {
   "isNotEmpty": "foobar should not be empty" 
  } 
}
```

or:

```
 { 
  "value": null, 
  "property": "foobar", 
  "children": [], 
  "constraints": [
    {"isNotEmpty": "foobar should not be empty" },
    {"isInt": "foobar must be an integer number"},
    { "min": "foobar must be greater than 0"}
  ]
}
```

**Comment by NoNameProvided** [comment]
Is this a feature request? What would be your ideal format?

**Comment by HonoluluHenk** [comment]
Yes, this is a feature request :)

The result with all validations as list would be nice:
```
{ "value": null,
  "property": "foobar",
  "children": [], 
  "constraints": [ 
    {"isNotEmpty": "foobar should not be empty" },
    {"isInt": "foobar must be an integer number"},
    { "min": "foobar must be greater than 0"}
}
```
with the constraints ordering same as annotation ordering

**Comment by kostanev** [comment]
It would be nice if there is a setting that makes the validator stop after the very first fail. For example, I have custom constraint that checks the database for email existence. I used the following decorators:
```
@IsNotEmpty()
@Length(3, 10)
@EmailNotExists()
email: string;
```
This makes database query even if email is empty or with less than 3 characters. Of course I could add additional checks in my constraint, but it makes no sense to do that.

**Comment by HonoluluHenk** [comment]
@kostanev is making a good point

**Comment by HonoluluHenk** [comment]
Regarding: making the validator stop after the very first fail...

I just had a look a Java Bean Validation: https://docs.oracle.com/javaee/6/tutorial/doc/gircz.html
(or better said: I've been jusing Bean Validation for years :)) and its behavior towards null just makes it easy to use:
all validations allow null, except those that explicitly check for null (and/or undefined for our case).

This in combination with validation in declaration-order make for a intuitive use... maybe this would be nice here, too? This would also solve kostanev's and my problem...

Also: @IsOptional could then be removed

**Comment by HonoluluHenk** [comment]
This null/undefined behaviour bugged me so much I just implemented this :)
See commit 4ac6797

**Comment by HonoluluHenk** [comment]
And validation with only 1 validation result is implemented in 1ee346b

**Comment by 19majkel94** [comment]
> all validations allow null, except those that explicitly check for null (and/or undefined for our case).
Also: @IsOptional could then be removed

Just use `skipMissingProperties: true` with `@IsDefined`

**Comment by ph55** [comment]
I agree.
Would be much efficient when no validators would run after one failed.

My example:
```ts
  @IsInt()
  @IsInModel({ model: User, property: 'id' })
  userId: number;
```

`IsInModel` will still run even though `userId` sent as string.

And error will look doubled and not so smart: `User id must be an integer number. Invalid user id`

**Comment by guoerwei** [comment]
+1
I want an option for this feature

**Comment by tiagocorreiaalmeida** [comment]
Any update on the topic? Just came around this, and I think it can become a performance issue to be requesting the database everytime a new request comes in with invalid data

**Comment by TranBaVinhSon** [comment]
Any update on this issue?

**Comment by cesc1802** [comment]
I also need a feature

**Comment by faiztherocker** [comment]
The feature to stop executing the validation rules if the first validation fails would help alot.

**Comment by Bezlepkin** [comment]
+1. I need too

**Comment by molinab297-unisys** [comment]
For those that are interested, I had the exact same problem as @ph55 . You can use validation groups to solve this issue: 
```javascript
  @IsInt({groups: ["validate-types"] })
  @IsInModel({ model: User, property: 'id' })
  userId: number;
```
and then run the validation twice:
```javascript
    await validateOrReject(this, {
      groups: ["validate-types"]
    });
    await validateOrReject(this);
```
The first validateOrReject call only runs the "type check" validators, and the second call runs through all of the validators. By doing this, you can avoid running validators on that values that don't match the type you expect (which as others have mentioned, makes no sense and is completely useless).

**Comment by molinab297-unisys** [comment]
Looks like #620 will actually fix this issue...

**Comment by NoNameProvided** [comment]
This has been fixed in develop via #620 and be included in the next release.
